### PR TITLE
add llama + sd models to iree test suite

### PR DESCRIPTION
**Pull request by saienduri**
This commit plugs in turbine tank exported mlir to the iree testing framework for llama + sd models.
Couldn't add real weight test for llama. Real weight file is 20GB and crashing the runner when trying.
On that note, looks like splat test cases aren't running at the moment?
Maybe we can enable the real weight test after we get the cluster.
Also, just in case it's useful for anyone adding models in future, for the exported model mlir, I had to update real weight flag to `--parameters=model=real_weights.irpa` to get it working.

### CODE REVIEW

File: iree_tests/pytorch/models/llama-tank/test_cases.json
Comment (ScottTodd) [inline iree_tests/pytorch/models/llama-tank/test_cases.json:5]: > On that note, looks like splat test cases aren't running at the moment?

The test code is filtering to just real weight tests (using `-k real_weights`) right now:

https://github.com/nod-ai/SHARK-TestSuite/blob/73dcb4239cea621e4d5366b04f2bde7ffccd4470/.github/workflows/test_iree.yml#L75-L78

I consider the `test_iree.yml` in this repo to be a sanity check and demonstration of the test suite itself, not a source of truth for test coverage. https://github.com/openxla/iree/pull/16801 adds similar tests in the IREE repo, and that is what "really matters" (will run on presubmit and should be kept passing). To that end, we can add whatever new filtering and test cases we want here.

File: iree_tests/pytorch/models/sd-unet-tank/test_cases.json
Comment (ScottTodd) [inline iree_tests/pytorch/models/sd-unet-tank/test_cases.json:10]: These tests timed out (and then were retried... which then ignored the timeout... that's an odd interaction, pytest plugins :P)

https://github.com/nod-ai/SHARK-TestSuite/actions/runs/8433834658/job/23095725951?pr=126#step:10:26

How about increasing the timeout here to 600 seconds? https://github.com/nod-ai/SHARK-TestSuite/blob/73dcb4239cea621e4d5366b04f2bde7ffccd4470/.github/workflows/test_iree.yml#L75-L78 (that's what I have it set to in https://github.com/openxla/iree/pull/16801 over in the IREE repo)

File: iree_tests/pytorch/models/sd-vae-decode-tank/test_cases.json
Comment (ScottTodd) [inline iree_tests/pytorch/models/sd-vae-decode-tank/test_cases.json:13]: FYI, I'm considering collapsing these back down to just URLs like
`https://sharkpublic.blob.core.windows.net/sharkpublic/sai/sd-vae-decode-tank/sd-vae-decode-tank.mlirbc`
I was trying to be fancy with Azure downloading and saving a few characters in .json files but making the URLs easy to copy/paste and `wget` is going to be nicer for developers

File: iree_tests/pytorch/models/llama-tank/real_weights_data_flags.txt
Comment (ScottTodd) [inline iree_tests/pytorch/models/llama-tank/real_weights_data_flags.txt:7]: Woah that's a lot of large outputs. Do we need to be checking all of them?

Is this coming from Turbine, or torchscript / "direct"?

**Review by ScottTodd — COMMENTED** [review]
Nice! Can you also review https://github.com/openxla/iree/pull/16801 please? Then once this lands we can start testing these programs on IREE presubmit too.

File: iree_tests/pytorch/models/llama-tank/real_weights_data_flags.txt
Comment (saienduri) [reply iree_tests/pytorch/models/llama-tank/real_weights_data_flags.txt]: This is coming from Turbine. These are just the tensor outputs of a llama forward pass (https://huggingface.co/docs/transformers/v4.39.1/en/main_classes/output#transformers.modeling_outputs.CausalLMOutputWithPast). Specifically, it is the logits and past_key_values tensors. We can ignore the past_key_value tensors and just look at logits. Let me know if I used the ignored outputs syntax correctly.

File: iree_tests/pytorch/models/llama-tank/test_cases.json
Comment (saienduri) [reply iree_tests/pytorch/models/llama-tank/test_cases.json:5]: Got it! Thanks for the info.

File: iree_tests/pytorch/models/sd-vae-decode-tank/test_cases.json
Comment (saienduri) [reply iree_tests/pytorch/models/sd-vae-decode-tank/test_cases.json:13]: Ok cool. This is also pretty nice though. Just being able to see the file names instead of the cluttered urls can be nice for anyone taking a look. Maybe we can add a comment that the base url to each file like: `https://sharkpublic.blob.core.windows.net/sharkpublic/sai/sd-vae-decode-tank` is the base url. If interested in downloading the file simply wget base url + `/{file_name}`

Comment (ScottTodd) [reply iree_tests/pytorch/models/sd-vae-decode-tank/test_cases.json:13]: Sure. I at least consider what's in the json files today to be very rough V0 code. So I wouldn't treat it as sacred if you find it tricky to work with.

Another ideas was to group files for each test case using tar/zip, so we wouldn't need to list them all explicitly.

File: iree_tests/pytorch/models/llama-tank/real_weights_data_flags.txt
Comment (ScottTodd) [reply iree_tests/pytorch/models/llama-tank/real_weights_data_flags.txt]: > Let me know if I used the ignored outputs syntax correctly.

You may need to repeat `--expected_output="(ignored)"` for each output (could try with a smaller test case to confirm)

File: iree_tests/pytorch/models/llama-tank/test_cases.json
Comment (ScottTodd) [reply iree_tests/pytorch/models/llama-tank/test_cases.json:5]: Speaking of how brittle the tests are in this repo (by using the latest IREE nightlies instead of a pinned version), you may want to sync past https://github.com/nod-ai/SHARK-TestSuite/commit/28091b0996239f411df5850b1a7e41875ae047d2 to get a clean CI run on this PR

File: iree_tests/pytorch/models/sd-vae-decode-tank/test_cases.json
Comment (saienduri) [reply iree_tests/pytorch/models/sd-vae-decode-tank/test_cases.json:13]: Oooh that would be much easier for adding the tests

File: iree_tests/pytorch/models/llama-tank/real_weights_data_flags.txt
Comment (saienduri) [reply iree_tests/pytorch/models/llama-tank/real_weights_data_flags.txt:7]: Tried a couple different variations (complains about no shape being provided and when I did, it just complained about being unable to parse arg). Maybe come back to it when we actually enable llama real weight test?

Comment (ScottTodd) [reply iree_tests/pytorch/models/llama-tank/real_weights_data_flags.txt:7]: ¯\\\_(ツ)\_/¯

> Maybe come back to it when we actually enable llama real weight test?

SGTM

Comment (saienduri) [reply iree_tests/pytorch/models/llama-tank/real_weights_data_flags.txt:7]: Cool, are we good to land this?

**Review by ScottTodd — APPROVED** [review]
Regarding naming, is "turbine tank" different from https://github.com/nod-ai/SHARK-TestSuite/tree/main/e2eshark/pytorch/models ? Is that https://github.com/nod-ai/SHARK-Turbine/tree/main/models/turbine_models/turbine_tank ?

Eventually we should have scripts like https://github.com/nod-ai/SHARK-TestSuite/blob/main/iree_tests/pytorch/models/import_from_e2eshark.py or guides like https://github.com/nod-ai/SHARK-TestSuite/tree/main/iree_tests#pytorch-models for each source of model files.

Just looking ahead a bit to whenever the MLIR and input/output files need to be regenerated. We want enough either automated or at least documented such that anyone can update the files*

\* if they have write access to the cloud projects, but some github actions secrets + automation can also help there too

**Comment by saienduri** [comment]
Yeah, it is different. I want to get that process fully automated (Pull in the latest passing mlir nightly). Not sure if we can have a guide for adding turbine tank models though because I had to mess around with the model runners itself, so that it generates the input/output .bin files. The rest of the process (converting mlir to mlirbc and safetensors to splats, real_weights irpas) seems to already be covered.

File: iree_tests/pytorch/models/sd-clip-tank/sd-clip-tank.mlirbc
Comment (ScottTodd) [inline iree_tests/pytorch/models/sd-clip-tank/sd-clip-tank.mlirbc:1]: Oh, this should have been checked in using git lfs, based on https://github.com/nod-ai/SHARK-TestSuite/blob/9464d503ddf37e658cf267254ac65a3d1f143344/iree_tests/.gitattributes#L1-L2

I see a warning when I pull this:
```
Encountered 6 file(s) that should have been pointers, but weren't:
        iree_tests/pytorch/models/llama-tank/splats.irpa
        iree_tests/pytorch/models/sd-clip-tank/sd-clip-tank.mlirbc
        iree_tests/pytorch/models/sd-clip-tank/splats.irpa
        iree_tests/pytorch/models/sd-unet-tank/sd-unet-tank.mlirbc
        iree_tests/pytorch/models/sd-unet-tank/splats.irpa
        iree_tests/pytorch/models/sd-vae-decode-tank/splats.irpa
```

I'm also learning how to use git lfs, so I guess I'll get back to you when I know how to resolve that and stop it from happening again...

Comment (ScottTodd) [reply iree_tests/pytorch/models/sd-clip-tank/sd-clip-tank.mlirbc:1]: Sent https://github.com/nod-ai/SHARK-TestSuite/pull/147 to fix

**Pull request merged**
